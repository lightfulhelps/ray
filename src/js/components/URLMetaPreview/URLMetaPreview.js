// @flow
import React from 'react';
import classNames from 'classnames';
import LinesEllipsis from 'react-lines-ellipsis';
import responsiveHOC from 'react-lines-ellipsis/lib/responsiveHOC';

type Props = {
  className?: string,
  description?: string,
  image?: string,
  isVertical?: boolean,
  title?: string,
  url?: string,
};

const cleanURL = (url: string): string => url.replace(/https?:\/\/(www.)?/, '');

const ResponsiveEllipsis = responsiveHOC()(LinesEllipsis);

const URLMetaPreview = ({
  className,
  description,
  image,
  isVertical = false,
  title,
  url,
  ...other
}: Props) => {
  const blockClass = 'url-meta-preview';
  const classes = classNames(className, blockClass);

  if (!url) {
    return null;
  }

  return (
    <div {...other} className={classes}>
      <div className={isVertical ? '' : 'row h-100'}>
        {image && (
          <div className={isVertical ? '' : 'col-4 pr-0'}>
            <div className={`${blockClass}__image`} style={{ backgroundImage: `url(${image})` }}>
              {isVertical && <img src={image} alt="" />}
            </div>
          </div>
        )}
        <div className={isVertical ? '' : image ? 'col-8 pl-0' : 'col-12'}>
          <div className={`${blockClass}__main`}>
            <a
              className={`${blockClass}__title`}
              href={url}
              target="_blank"
              rel="noopener noreferrer"
            >
              <ResponsiveEllipsis text={title} maxLine={2} ellipsis="..." basedOn="words" />
            </a>
            {description && (
              <ResponsiveEllipsis
                className={`${blockClass}__description`}
                text={description}
                maxLine={3}
                ellipsis="..."
                basedOn="words"
              />
            )}
            <a
              className={`${blockClass}__url`}
              href={url}
              target="_blank"
              rel="noopener noreferrer"
            >
              {cleanURL(url)}
            </a>
          </div>
        </div>
      </div>
    </div>
  );
};

export default URLMetaPreview;
