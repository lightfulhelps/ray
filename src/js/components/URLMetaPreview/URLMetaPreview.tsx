import React from 'react';
import classNames from 'classnames';

type Props = {
  className?: string;
  description?: string;
  image?: string;
  isVertical?: boolean;
  title?: string;
  url?: string;
};

const cleanURL = (url: string): string => url.replace(/https?:\/\/(www.)?/, '');

const URLMetaPreview: React.FC<Props> = ({
  className,
  description,
  image,
  isVertical = false,
  title,
  url,
  ...other
}: Props) => {
  const classes = classNames(className, 'url-meta-preview', 'bg-white h-100');

  if (!url) {
    return null;
  }

  return (
    <div {...other} className={classes}>
      <div className={isVertical ? '' : 'row h-100'}>
        {image && (
          <div className={isVertical ? '' : 'col-4 pr-0'}>
            <div
              className="w-100 h-100 bg-cover bg-center"
              data-test-id="url-meta-preview-image"
              style={{ backgroundImage: `url(${image})` }}
            >
              {isVertical && <img className="w-100" src={image} alt="" />}
            </div>
          </div>
        )}
        <div className={isVertical ? '' : image ? 'col-8 pl-0' : 'col-12'}>
          <div className="d-flex flex-column justify-content-between h-100 p-1 text-xs">
            <a
              className="font-weight-bold text-dark text-truncate"
              href={url}
              target="_blank"
              rel="noopener noreferrer"
            >
              {title}
            </a>
            {description && (
              <div
                className="font-weight-light text-gray-500 text-truncate"
                data-test-id="url-meta-preview-description"
              >
                {description}
              </div>
            ) // <ResponsiveEllipsis
            //   className="font-weight-light text-gray-500"
            //   text={description}
            //   maxLine={3}
            //   ellipsis="..."
            //   basedOn="words"
            // />
            }
            <a
              className="d-block text-body text-truncate"
              data-test-id="url-meta-preview-url"
              href={url}
              target="_blank"
              rel="noopener noreferrer"
            >
              {cleanURL(url)}
            </a>
          </div>
        </div>
      </div>
    </div>
  );
};

export default URLMetaPreview;
