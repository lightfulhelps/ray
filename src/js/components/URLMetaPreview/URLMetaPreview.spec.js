import React from 'react';
import { shallow } from 'enzyme';
import URLMetaPreview from './URLMetaPreview';

const setup = (overrides = {}) => {
  const props = {
    title: "National Living Wage 'fails to cover families' basic needs",
    description:
      "Low-paid parents working full-time do not earn enough to meet their families' needs, a charity says.",
    image:
      'https://ichef.bbci.co.uk/news/1024/branded_news/81C6/production/_98022233_gettyimages-842414672-1.jpg',
    url: 'https://www.bbc.co.uk/news/business-45242008',
    ...overrides,
  };
  const wrapper = shallow(<URLMetaPreview {...props} />);

  return {
    wrapper,
    props,
  };
};

describe('<URLMetaPreview />', () => {
  it('should render', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should render in vertical mode', () => {
    const { wrapper } = setup({ isVertical: true });

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle the className prop', () => {
    const { wrapper } = setup();

    wrapper.setProps({ className: 'custom' });

    expect(wrapper.hasClass('url-meta-preview')).toBe(true);
    expect(wrapper.hasClass('custom')).toBe(true);
  });

  it('should handle no image', () => {
    const { wrapper } = setup();

    expect(wrapper.find('.url-meta-preview__image').exists()).toBe(true);

    wrapper.setProps({ image: null });

    expect(wrapper.find('.url-meta-preview__image').exists()).toBe(false);
  });

  it('should handle no description', () => {
    const { wrapper } = setup();

    expect(wrapper.find('.url-meta-preview__description').exists()).toBe(true);

    wrapper.setProps({ description: null });

    expect(wrapper.find('.url-meta-preview__description').exists()).toBe(false);
  });

  it('should format the url', () => {
    const { wrapper } = setup();

    expect(wrapper.find('.url-meta-preview__url').text()).toBe('bbc.co.uk/news/business-45242008');
  });
});
