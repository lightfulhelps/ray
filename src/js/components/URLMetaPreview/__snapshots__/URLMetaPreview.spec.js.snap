// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<URLMetaPreview /> should render 1`] = `
<div
  className="url-meta-preview"
>
  <div
    className="row h-100"
  >
    <div
      className="col-4 pr-0"
    >
      <div
        className="url-meta-preview__image"
        style={
          Object {
            "backgroundImage": "url(https://ichef.bbci.co.uk/news/1024/branded_news/81C6/production/_98022233_gettyimages-842414672-1.jpg)",
          }
        }
      />
    </div>
    <div
      className="col-8 pl-0"
    >
      <div
        className="url-meta-preview__main"
      >
        <a
          className="url-meta-preview__title"
          href="https://www.bbc.co.uk/news/business-45242008"
          rel="noopener noreferrer"
          target="_blank"
        >
          <Responsive(LinesEllipsis)
            basedOn="words"
            ellipsis="..."
            innerRef={[Function]}
            maxLine={2}
            text="National Living Wage 'fails to cover families' basic needs"
          />
        </a>
        <Responsive(LinesEllipsis)
          basedOn="words"
          className="url-meta-preview__description"
          ellipsis="..."
          innerRef={[Function]}
          maxLine={3}
          text="Low-paid parents working full-time do not earn enough to meet their families' needs, a charity says."
        />
        <a
          className="url-meta-preview__url"
          href="https://www.bbc.co.uk/news/business-45242008"
          rel="noopener noreferrer"
          target="_blank"
        >
          bbc.co.uk/news/business-45242008
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`<URLMetaPreview /> should render in vertical mode 1`] = `
<div
  className="url-meta-preview"
>
  <div
    className=""
  >
    <div
      className=""
    >
      <div
        className="url-meta-preview__image"
        style={
          Object {
            "backgroundImage": "url(https://ichef.bbci.co.uk/news/1024/branded_news/81C6/production/_98022233_gettyimages-842414672-1.jpg)",
          }
        }
      >
        <img
          alt=""
          src="https://ichef.bbci.co.uk/news/1024/branded_news/81C6/production/_98022233_gettyimages-842414672-1.jpg"
        />
      </div>
    </div>
    <div
      className=""
    >
      <div
        className="url-meta-preview__main"
      >
        <a
          className="url-meta-preview__title"
          href="https://www.bbc.co.uk/news/business-45242008"
          rel="noopener noreferrer"
          target="_blank"
        >
          <Responsive(LinesEllipsis)
            basedOn="words"
            ellipsis="..."
            innerRef={[Function]}
            maxLine={2}
            text="National Living Wage 'fails to cover families' basic needs"
          />
        </a>
        <Responsive(LinesEllipsis)
          basedOn="words"
          className="url-meta-preview__description"
          ellipsis="..."
          innerRef={[Function]}
          maxLine={3}
          text="Low-paid parents working full-time do not earn enough to meet their families' needs, a charity says."
        />
        <a
          className="url-meta-preview__url"
          href="https://www.bbc.co.uk/news/business-45242008"
          rel="noopener noreferrer"
          target="_blank"
        >
          bbc.co.uk/news/business-45242008
        </a>
      </div>
    </div>
  </div>
</div>
`;
